s3_reader.go

```diff
@@ -0,0 +1,101 @@
+package main
+
+import (
+	"bytes"
+	"log"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+)
+
+type S3ReadCloser struct {
+	bucket    string
+	key       string
+	offset    int
+	totalSize int
+	readBytes int
+	sess      *session.Session
+	buf       *bytes.Buffer
+}
+
+func awsConfig() *aws.Config {
+	region := os.Getenv("AWS_DEFAULT_REGION")
+	if region == "" {
+		region = os.Getenv("AWS_REGION")
+		if region == "" {
+			region = "us-east-1"
+		}
+	}
+
+	config := &aws.Config{Region: aws.String(region)}
+
+	if endpoint := os.Getenv("AWS_ENDPOINT_URL"); endpoint != "" {
+		config.Endpoint = aws.String(endpoint)
+		log.Println("Custom endpoint:", endpoint)
+	}
+
+	log.Println("Connecting to S3. Region: " + region)
+
+	config.CredentialsChainVerboseErrors = aws.Bool(true)
+
+	if os.Getenv("AWS_DEBUG") != "" {
+		config.LogLevel = aws.LogLevel(aws.LogDebugWithHTTPBody)
+	}
+
+	return config
+}
+
+func NewS3ReadCloser(path string) *S3ReadCloser {
+	if !PRO {
+		log.Fatal("Using S3 input and output require PRO license")
+		return nil
+	}
+
+	bucket, key := parseS3Url(path)
+	sess := session.Must(session.NewSession(awsConfig()))
+
+	log.Println("[S3 Input] S3 connection succesfully initialized", path)
+
+	return &S3ReadCloser{
+		bucket: bucket,
+		key:    key,
+		sess:   sess,
+		buf:    &bytes.Buffer{},
+	}
+}
+
+func (s *S3ReadCloser) Read(b []byte) (n int, e error) {
+	if s.readBytes == 0 || s.readBytes+len(b) > s.offset {
+		svc := s3.New(s.sess)
```

Comment:
we really have to create new session on each read operation?

Reply:
more like merging former configuration into the newly created session https://github.com/aws/aws-sdk-go/blob/master/aws/session/session.go#L53